staging/src/k8s.io/apiextensions-apiserver/pkg/controller/openapi/builder.go

```diff
@@ -322,6 +324,42 @@ func (b *builder) buildKubeNative(schema *structuralschema.Structural, v2 bool)
 	return ret
 }
 
+func addEmbeddedProperties(s *spec.Schema) {
+	if s == nil {
+		return
+	}
+	for k := range s.Properties {
+		v := s.Properties[k]
+		addEmbeddedProperties(&v)
+		s.Properties[k] = v
+	}
+	if s.Items != nil {
+		addEmbeddedProperties(s.Items.Schema)
+	}
+	if isTrue, ok := s.VendorExtensible.Extensions.GetBool("x-kubernetes-embedded-resource"); ok && isTrue {
+		addTypeMetaProperties(s)
+
+		req := sets.NewString(s.Required...)
+
+		if !req.Has("kind") {
+			s.Required = append(s.Required, "kind")
```

Comment:
we don't require kind and apiVersion in builtin schema. We can remove the requirement here to be consistent. It's okay for client-side validation to be less strict

Reply:
RawExtensions need it. Otherwise, it is not a valid object.

Reply:
I assume we require them in server-side validation, and that will be sufficient

having the requirement in client-side doesn't hurt, perhaps we could add the same to the [builtin schemata](https://raw.githubusercontent.com/kubernetes/kubernetes/master/api/openapi-spec/swagger.json) later to be consistent

Reply:
It's required for embedded. In some cases, I think it can be inferred from the endpoint for the top-level resources. I would still like to make it required from the beginning for embedded resources here

Reply:
+1

Reply:
required now.